src/backend/gl/gl_common.c

```diff
@@ -1154,10 +1162,33 @@ void gl_present(backend_t *base, const region_t *region) {
 	glDeleteBuffers(2, bo);
 	glDeleteVertexArrays(1, &vao);
 
+	glEndQuery(GL_TIME_ELAPSED);
+	gd->current_frame_timing ^= 1;
+
+	gl_check_err();
+
 	free(coord);
 	free(indices);
 }
 
+bool gl_last_render_time(backend_t *base, struct timespec *ts) {
+	auto gd = (struct gl_data *)base;
+	GLint available = 0;
+	glGetQueryObjectiv(gd->frame_timing[gd->current_frame_timing ^ 1],
+	                   GL_QUERY_RESULT_AVAILABLE, &available);
+	if (!available) {
+		return false;
+	}
+
+	GLuint64 time;
+	glGetQueryObjectui64v(gd->frame_timing[gd->current_frame_timing ^ 1],
```

Comment:
a note from https://www.khronos.org/opengl/wiki/Query_Object: `Note that implementations are allowed to return 0 for any of these values. This means that the query type contains no actual information, and the implementation does not really support it.`

however, i don't think that it's possible to meet such a funny opengl implementation in the real word, so let it be just a note.